homeassistant/components/husqvarna_automower_ble/coordinator.py

```diff
@@ -0,0 +1,100 @@
+from bleak import BleakError
+from bleak_retry_connector import close_stale_connections_by_address
+
+from homeassistant.components import bluetooth
+from homeassistant.core import HomeAssistant
+from homeassistant.helpers.update_coordinator import DataUpdateCoordinator, UpdateFailed
+
+from .const import DOMAIN, LOGGER
+
+SCAN_INTERVAL = timedelta(seconds=60)
+
+
+class HusqvarnaCoordinator(DataUpdateCoordinator[dict[str, bytes]]):
+    """Class to manage fetching data."""
+
+    def __init__(
+        self,
+        hass: HomeAssistant,
+        mower: Mower,
+        address: str,
+        channel_id: str,
+        model: str,
+    ) -> None:
+        """Initialize global data updater."""
+        super().__init__(
+            hass=hass,
+            logger=LOGGER,
+            name=DOMAIN,
+            update_interval=SCAN_INTERVAL,
+        )
+        self.address = address
+        self.channel_id = channel_id
+        self.model = model
+        self.mower = mower
+
+    async def async_shutdown(self) -> None:
+        """Shutdown coordinator and any connection."""
+        LOGGER.debug("Shutdown")
+        await super().async_shutdown()
+        if self.mower.is_connected():
+            await self.mower.disconnect()
+
+    async def _async_find_device(self):
+        LOGGER.debug("Trying to reconnect")
+        await close_stale_connections_by_address(self.address)
+
+        device = bluetooth.async_ble_device_from_address(
+            self.hass, self.address, connectable=True
+        )
+
+        try:
+            if not await self.mower.connect(device):
+                raise UpdateFailed("Failed to connect")
+        except BleakError as err:
+            raise UpdateFailed("Failed to connect") from err
+
+    async def _async_update_data(self) -> dict[str, bytes]:
+        """Poll the device."""
+        LOGGER.debug("Polling device")
+
+        data: dict[str, bytes] = {}
+
+        try:
+            if not self.mower.is_connected():
+                await self._async_find_device()
+        except BleakError as err:
+            raise UpdateFailed("Failed to connect") from err
+
+        try:
+            data["battery_level"] = await self.mower.battery_level()
+            LOGGER.debug(data["battery_level"])
+            if data["battery_level"] is None:
+                await self._async_find_device()
+                raise UpdateFailed("Error getting data from device")
+
+            data["activity"] = await self.mower.mower_activity()
+            LOGGER.debug(data["activity"])
+            if data["activity"] is None:
+                await self._async_find_device()
+                raise UpdateFailed("Error getting data from device")
+
+            data["state"] = await self.mower.mower_state()
+            LOGGER.debug(data["state"])
+            if data["state"] is None:
+                await self._async_find_device()
+                raise UpdateFailed("Error getting data from device")
```

Comment:
A update involves sending a request to the mower for every type of data (currently battery, state and activity) and waiting for a response. If the connection is lost (because the mower moves out of range or stops responding) then it's possible to only get some of the data. In reality it's not uncommon to happen.

So this is trying to recover the device if any of the data queries fail. If the mower if back in range then that's great, which is what we are trying to catch.

The other issue is that the mower can sometimes just stop responding for a little bit, so we are also trying to work around that.

Reply:
Would it be viable to merge them all? Because they all raise UpdateFailed, and if that happens, all coordinator entities are marked unavailable

Reply:
The idea is to error as soon as we can't get a single query. If we consolidate them all then we have to wait until the end to retry connecting. Which just uses more time

Reply:
Right